File: Archive_LJ_micro/gmdlj.f90
!**********************************************************************!
!
! File: gmdlj.f90
!
! Create random ("gas") initial configuration for NVE Molecular
! Dynamics of Lennard-Jonesium
!
! 10-May-1997 (MN)
! 04-May-2012
!
!**********************************************************************!

program gmdlj

!**********************************************************************!

  implicit none

  integer,parameter::iout=2

  character(len=80)::fname
  integer::i,m,n,ncor,nt,ntaskip,ntcskip,ntjob,ntorig,ntprint
  real::c,dr,dt,ran,rho
  real,dimension(:),allocatable::vx,vy,vz,x,y,z

! User input

  magic: do
    write(unit=*,fmt="(a)",advance="no") "               n="
    read(unit=*,fmt=*) n
    m=0
    do
      m=m+2
      if(m**3==2*n) then       ! Check if magic number
        exit magic
      else if(m**3>2*n) then
        exit
      end if
    end do
  end do magic

  write(unit=*,fmt="(a)",advance="no") "             rho="
  read(unit=*,fmt=*) rho
  write(unit=*,fmt="(a)",advance="no") "              dt="
  read(unit=*,fmt=*) dt
  write(unit=*,fmt="(a)",advance="no") "              dr="
  read(unit=*,fmt=*) dr
  write(unit=*,fmt="(a)",advance="no") "         ntaskip="
  read(unit=*,fmt=*) ntaskip
  write(unit=*,fmt="(a)",advance="no") " ntprint/ntaskip="
  read(unit=*,fmt=*) ntprint
  write(unit=*,fmt="(a)",advance="no") "         ntcskip="
  read(unit=*,fmt=*) ntcskip
  write(unit=*,fmt="(a)",advance="no") "    ncor/ntcskip="
  read(unit=*,fmt=*) ncor
  write(unit=*,fmt="(a)",advance="no") "  ntorig/ntcskip="
  read(unit=*,fmt=*) ntorig
  write(unit=*,fmt="(a)",advance="no") "           ntjob="
  read(unit=*,fmt=*) ntjob

  write(unit=*,fmt="(a)",advance="no") "           fname=[mdlj_in.dat] "
  read(unit=*,fmt="(a)") fname
  if(fname=="") then
    fname="mdlj_in.dat"
  end if

! Allocate arrays

  allocate(vx(n),vy(n),vz(n),x(n),y(n),z(n))

! Random positions

  c=(n/rho)**(1.0/3.0)
  do i=1,n
    call random_number(ran)
    x(i)=c*(ran-0.5)
    call random_number(ran)
    y(i)=c*(ran-0.5)
    call random_number(ran)
    z(i)=c*(ran-0.5)
  end do

! Zero velocities

  vx=0.0
  vy=0.0
  vz=0.0

! Write startup file

  nt=0

  open(unit=iout,file=fname,status="replace",action="write", &
       form="unformatted")

  write(unit=iout) n,ncor,nt,ntaskip,ntcskip,ntjob,ntorig,ntprint
  write(unit=iout) dr,dt,rho
  write(unit=iout) x,y,z
  write(unit=iout) vx,vy,vz

  close(unit=iout)

! Deallocate arrays

  deallocate(vx,vy,vz,x,y,z)

end program gmdlj

!**********************************************************************!

File: Archive_LJ_micro/gmdlj.c
/**********************************************************************
 *
 * File: gmdlj.c
 *
 * Create random ("gas") initial configuration for NVE Molecular
 * Dynamics of Lennard-Jonesium
 *
 * 04-May-2010 (MN)
 * 04-May-2012
 *
 **********************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#include "getval.h"

#define BSIZE 80

/**********************************************************************/

int main() {

/**********************************************************************/

  FILE *fpo;
  char fname[BSIZE];
  int n,ncor,nt,ntaskip,ntcskip,ntjob,ntorig,ntprint;
  int i,m;
  double dr,dt,rho;
  double c;
  double *vx,*vy,*vz,*x,*y,*z;

  /* User input */

  for(;;) {
    printf("               n=");
    getval_i(&n);
    m=0;
    while(m*m*m<2*n) {   /* Check if magic number */
      m=m+2;
      if(m*m*m==2*n) {
	goto magic;
      }
    }
  }
  magic:

  printf("             rho=");
  getval_d(&rho);
  printf("              dt=");
  getval_d(&dt);
  printf("              dr=");
  getval_d(&dr);
  printf("         ntaskip=");
  getval_i(&ntaskip);
  printf(" ntprint/ntaskip=");
  getval_i(&ntprint);
  printf("         ntcskip=");
  getval_i(&ntcskip);
  printf("    ncor/ntcskip=");
  getval_i(&ncor);
  printf("  ntorig/ntcskip=");
  getval_i(&ntorig);
  printf("           ntjob=");
  getval_i(&ntjob);

  strcpy(fname,"mdlj_in.dat");
  printf("           fname=[mdlj_in.dat] ");
  getval_s(fname);

  /* Allocate arrays */

  x=(double*)malloc(n*sizeof(double));
  y=(double*)malloc(n*sizeof(double));
  z=(double*)malloc(n*sizeof(double));
  vx=(double*)malloc(n*sizeof(double));
  vy=(double*)malloc(n*sizeof(double));
  vz=(double*)malloc(n*sizeof(double));

  /* Random positions & zero velocities */

  c=cbrt(n/rho);
  for(i=0;i<=n-1;i++) {
    x[i]=c*(drand48()-0.5);
    y[i]=c*(drand48()-0.5);
    z[i]=c*(drand48()-0.5);
    vx[i]=0.0;
    vy[i]=0.0;
    vz[i]=0.0;
  }

  /* Write startup file */

  nt=0;

  fpo=fopen(fname,"w");

  fwrite(&n,sizeof(int),1,fpo);
  fwrite(&ncor,sizeof(int),1,fpo);
  fwrite(&nt,sizeof(int),1,fpo);
  fwrite(&ntaskip,sizeof(int),1,fpo);
  fwrite(&ntcskip,sizeof(int),1,fpo);
  fwrite(&ntjob,sizeof(int),1,fpo);
  fwrite(&ntorig,sizeof(int),1,fpo);
  fwrite(&ntprint,sizeof(int),1,fpo);

  fwrite(&dr,sizeof(double),1,fpo);
  fwrite(&dt,sizeof(double),1,fpo);
  fwrite(&rho,sizeof(double),1,fpo);

  fwrite(x,sizeof(double),n,fpo);
  fwrite(y,sizeof(double),n,fpo);
  fwrite(z,sizeof(double),n,fpo);

  fwrite(vx,sizeof(double),n,fpo);
  fwrite(vy,sizeof(double),n,fpo);
  fwrite(vz,sizeof(double),n,fpo);

  fclose(fpo);

  /* Deallocate arrays */

  free(vx);
  free(vy);
  free(vz);
  free(x);
  free(y);
  free(z);

  return 0;

}

/**********************************************************************/
